### PR TITLE
ISPN-12908 Remove flag DONT_BUNDLE and fix parallel delivery

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.redhat.com/browse/ISPN-12908

### CODE REVIEW

**Comment by gustavolira** [comment]
@danberindei http://perfrepo.mw.lab.eng.bos.redhat.com/reports/tableComparisonReport/19138

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java
Comment (pruivo) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java:1487]: looking at JGroups code, `message` can be `null` in `forEach()´. it is probably worth checking for `null`.

Comment (danberindei) [reply core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java:1490]: I didn't add a check here, as I'm assuming each batch must have at least one non-null message, but I added a check at the beginning of the lambda (line 1472)

**Comment by danberindei** [comment]
I added the null check and one more commit with BlockHound exceptions.

**Review by pruivo — APPROVED** [review]

**Comment by pruivo** [comment]
@danberindei you have checkstyle errors in CI :)

**Comment by danberindei** [comment]
@pruivo I have no idea what possessed IntelliJ to change those imports, after reverting both Ctrl+Alt+L and Ctrl+Alt+O say the imports don't need to be changed. And I most definitely didn't add the `*` imports by hand :)

Doesn't really matter though, in the meantime I noticed a fairly reproducible failure in `XSiteAutoStateTransferTest` that I need to track down.

**Comment by danberindei** [comment]
Never mind, the `XSiteAutoStateTransferTest` failure also appears in master so I created https://issues.redhat.com/browse/ISPN-13012 for it

**Comment by danberindei** [comment]
run performance tests please
library cs

**Review by pruivo — APPROVED** [review]

**Comment by danberindei** [comment]
I put it on hold because the repl write throughput went down 6.68% with 10 threads (5981 writes/s), and the throughput with 100 threads is even worse (5358 writes/s). I will try to get some new results after the ISPN-12823 revert and perhaps give up on processing the messages in parallel.

**Comment by danberindei** [comment]
run performance tests please
library cs

**Comment by pruivo** [comment]
@danberindei can this be merged? or the `fix parallel delivery` part extracted to another PR?

**Comment by danberindei** [comment]
run performance tests please
library cs

**Comment by danberindei** [comment]
run performance tests please

**Comment by danberindei** [comment]
I could extract that part to a separate PR, but it's not going to make any difference in practice: as long as the DONT_BUNDLE flag is set, the target node is going to deliver each message separately. I've triggered another perf ack build and I'll try to look into the Library Repl Non-TX writes 100 threads throughput drop in order to get unblock this.

**Comment by danberindei** [comment]
run performance tests please

**Comment by danberindei** [comment]
run performance tests please

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java
Comment (pruivo) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java:1554]: What about using `org.infinispan.util.concurrent.BlockingManager#limitedBlockingExecutor()` with `concurrency=1` for the regular messages?

Comment (danberindei) [reply core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java:1553]: I'm not sure that would help... first off, the delivery should happen on a non-blocking thread, because everything but the sending of the reply should be non-blocking.

Second, UNICAST3/NAKACK2 use a Table as a queue for regular messages, which are only delivered after all previous messages have been delivered. If we pretend the regular message has been delivered and we put it in a limited executor's queue instead, it allows JGroups to remove the message faster, but I don't think it reduces the amount of work done by UNICAST3/NAKACK2.

PS: I already forgot I said I'd wait for the perf test results and I already pushed a change to process all messages in the delivery thread, not just regular messages :)

**Comment by danberindei** [comment]
Performance test result (d80632119b0d731ce15f705bf8ad1ea4343bb3ad used to be ISPN-12908 Fix tests before rebase):
http://perfrepo.mw.lab.eng.bos.redhat.com/reports/tableComparisonReport/19528

**Comment by pruivo** [comment]
perf looks ok to me. can you squash and remove the `sleep`? :) :)

**Comment by danberindei** [comment]
Done, I had forgotten about my early test failure reproduction attempts :)

**Comment by pruivo** [comment]
rebased and fixed conflicts in https://github.com/infinispan/infinispan/pull/10019

**Pull request closed without merging**
